pkg/es/cat_shards.go

```diff
@@ -0,0 +1,237 @@
+func (s *CatShardsService) Bytes(bytes string) *CatShardsService {
+	s.bytes = bytes
+	return s
+}
+
+// Local indicates to return local information, i.e. do not retrieve
+// the state from master node (default: false).
+func (s *CatShardsService) Local(local bool) *CatShardsService {
+	s.local = &local
+	return s
+}
+
+// MasterTimeout is the explicit operation timeout for connection to master node.
+func (s *CatShardsService) MasterTimeout(masterTimeout string) *CatShardsService {
+	s.masterTimeout = masterTimeout
+	return s
+}
+
+// Columns to return in the response.
+// To get a list of all possible columns to return, run the following command
+// in your terminal:
+//
+// Example:
+//   curl 'http://localhost:9200/_cat/shards?help'
+//
+// Please use the long names for columns (i.e. `completion.size`) for JSON unmarshalling
+// to work correctly.
+// You can use Columns("*") to return all possible columns. That might take
+// a little longer than the default set of columns.
+func (s *CatShardsService) Columns(columns ...string) *CatShardsService {
+	s.columns = columns
+	return s
+}
+
+// Sort is a list of fields to sort by.
+func (s *CatShardsService) Sort(fields ...string) *CatShardsService {
+	s.sort = fields
+	return s
+}
+
+// Pretty indicates that the JSON response be indented and human readable.
+func (s *CatShardsService) Pretty(pretty bool) *CatShardsService {
+	s.pretty = pretty
+	return s
+}
+
+// buildURL builds the URL for the operation.
+func (s *CatShardsService) buildURL() (string, url.Values, error) {
+	// Build URL
+	var (
+		path string
+		err  error
+	)
+
+	if s.index != "" {
+		path, err = uritemplates.Expand("/_cat/shards/{index}", map[string]string{
+			"index": s.index,
+		})
+	} else {
+		path = "/_cat/shards"
+	}
+	if err != nil {
+		return "", url.Values{}, err
+	}
+
+	// Add query string parameters
+	params := url.Values{
+		"format": []string{"json"}, // always returns as JSON
+	}
+	if s.pretty {
+		params.Set("pretty", "true")
+	}
+	if s.bytes != "" {
+		params.Set("bytes", s.bytes)
+	}
+	if v := s.local; v != nil {
+		params.Set("local", fmt.Sprint(*v))
+	}
+	if s.masterTimeout != "" {
+		params.Set("master_timeout", s.masterTimeout)
+	}
+	if len(s.columns) > 0 {
+		params.Set("h", strings.Join(s.columns, ","))
+	}
+	if len(s.sort) > 0 {
+		params.Set("s", strings.Join(s.sort, ","))
+	}
+	return path, params, nil
+}
+
+// Do executes the operation.
+func (s *CatShardsService) Do(ctx context.Context) (CatShardsResponse, error) {
+	// Get URL for request
+	path, params, err := s.buildURL()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get HTTP response
+	res, err := s.client.PerformRequest(ctx, elastic.PerformRequestOptions{
+		Method: "GET",
+		Path:   path,
+		Params: params,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Return operation response
+	var ret CatShardsResponse
+	if err := (&elastic.DefaultDecoder{}).Decode(res.Body, &ret); err != nil {
+		return nil, err
+	}
+	return ret, nil
```

Comment:
Could this function be abstracted away? It seems very generic. Although, it might make sense in Go in a way I'm unaware.

Reply:
I'm not entirely sure what you mean here? The `Do()` method? Abstracted how?

Everything in `pkg/es` follows the style of https://github.com/olivere/elastic. I'd like to contribute the rest of these if he ever gets around to merging my initial [pull request](https://github.com/olivere/elastic/pull/1101).

Reply:
Ah, that makes sense. 👍